Platform improvements

danielsreichenbach opened the pull request:
# Improvements made

This pull request modernizes the TSWoW framework build system, addresses security vulnerabilities, and improves TypeScript compatibility.

## Build System Changes

- Migrated from ts-node to SWC compiler for faster builds
- Updated TypeScript from 4.x to 5.6.2 with compatibility fixes across 73 files
- Fixed module resolution in tsconfig.json files throughout the codebase
- Updated npm dependencies including mysql2 and Mocha

## Security and Stability

- Fixed memory leaks in TSSmartScript.cpp, TSWorldPacket.cpp, and TSCustomPacket.cpp
- Added bounds validation for packet and array operations
- Implemented 8MB packet size limit to prevent denial of service
- Added SQL injection warnings to raw query functions

## Database Improvements

- Fixed MySQL connection pooling to respect configuration settings
- Added 10-second connection timeout handling
- Removed MySQL2 acquireTimeout warnings for non-pooled connections

## Developer Experience

- Added `--debug` parameter for datascripts and livescripts builds
- Extended CLI with `clean all` and `exit` commands
- Implemented quiet output by default with verbose logging via `--debug`
- Added Mocha testing framework with test templates

## New Files Added

- Configuration: `.swcrc`, `.mocharc.json`, `mise.toml`, `build-config.js`
- Testing: Test templates and utilities in `misc/install-config/test-template/`
- Core utilities: `TSRegExp.h`, `TSUtils.h`, `BuildConfig.ts`

## Modified Components

- tswow-core: 15 files updated for memory safety and type definitions
- tswow-scripts: 45 files updated for TypeScript 5.6 compatibility
- Build scripts: Updated for SWC integration and improved error handling

No breaking changes. All updates maintain backward compatibility.

## Dependencies

In order to fully work this, PR depends on

- [adt-creator Linux fixes](https://github.com/tswow/adt-creator/pull/1) to be merged
- adt-creator submodule updated to the latest "master" branch after merging above PR
- TrinityCore submodule updated to the latest "tswow"

These would be tasks left up to the project maintainers :smile:

danielsreichenbach committed:
chore: add project configuration and coding standards

- Add .editorconfig for code formatting across editors
- Add .markdownlint.yaml for markdown documentation standards
- Add .npmrc for npm behavior and settings
- Add mise.toml for tool version management
- Update .gitignore with patterns and build artifacts

These configuration files establish development standards
and tooling across the project.

danielsreichenbach committed:
ci: update GitHub Actions to Ubuntu 24.04 and Node.js 22

- Update Ubuntu runner from 22.04 to 24.04
- Update Node.js from 18.12.1 to 22
- Update GitHub Actions versions:
  - actions/checkout@v2 → v4
  - actions/setup-node@v3 → v4
- Package dependencies remain compatible with Ubuntu 24.04

Updates CI/CD infrastructure and avoids deprecation warnings
from GitHub Actions.

danielsreichenbach committed:
build: migrate from ts-node to SWC

- Replace ts-node with SWC for TypeScript compilation
- Add .swcrc configuration files for TypeScript handling
- Update build.js to use SWC for bootstrapping
- Add build-config.js for configuration management
- Add BuildConfig.ts utility for accessing build configuration
- Update CompileTsWow.ts for SWC compatibility
- Remove dependency on tsc-watch and ts-node

Reduces build times and startup times while maintaining
TypeScript compatibility.

danielsreichenbach committed:
test: add testing infrastructure with Mocha

- Add .mocharc.json for test configuration
- Add test templates and utilities in misc/install-config/test-template/
- Add Tests.ts command for running tests
- Add TestUtils.ts and TestsDirectory.ts for test utilities
- Configure Mocha with TypeScript support

Provides testing framework for the project with configuration
and utilities for writing and running tests.

danielsreichenbach committed:
deps: update npm dependencies

- Update TypeScript to 5.6.2 (from 4.x)
- Update mysql2 to 3.14.3 with TypeScript types
- Update Mocha to 11.7.1
- Add @swc/core and @swc/cli for build system
- Update terminal-kit, js-yaml, xml2js, fs-extra
- Fix deprecated package warnings
- Add missing @types/crypto-js dependency
- Remove deprecated ts-node and tsc-watch

Updates dependencies and resolves deprecation warnings
and security vulnerabilities.

danielsreichenbach committed:
fix: MySQL connection handling and pooling configuration

- Fix connection pooling to respect NodeConfig.UsePooling
- Remove MySQL2 acquireTimeout warnings for non-pooled connections
- Separate pool-specific options from connection options
- Add connection timeout handling (10 seconds)
- Update error messages for connection failures
- Remove conflicting connection settings

Resolves MySQL connection warnings and updates connection
error handling.

danielsreichenbach committed:
feat: add debug mode and new commands

- Add --debug parameter to datascripts and livescripts builds
- Hide verbose output by default, show details with --debug
- Add 'clean all' command alongside 'build all'
- Add 'exit' command for REPL shutdown
- Update command help descriptions and documentation
- Add error messages and debugging capabilities

Reduces build output noise while maintaining debugging
capabilities.

danielsreichenbach committed:
fix: code quality and security in tswow-core

Memory Management Fixes:
- Fix memory leak in TSSmartScript using stack allocation
- Fix TSWorldPacket destructor memory leak with Rule of Five
- Fix TSCustomPacket memory leak using RAII with shared_ptr

Security Fixes:
- Add bounds checking to TSWorldPacket operations
- Add bounds checking to TSArray index operations
- Add packet size validation (8MB limit) to prevent DoS
- Add SQL injection warnings to raw query functions

Type Safety Fixes:
- Fix type mismatches in TSString operations (double vs size_t)
- Add bounds validation for string operations
- Add explicit type casts to prevent conversion issues

Additional Changes:
- Document TSMutex copy constructor behavior
- Add TSRegExp.h and TSUtils.h utilities
- Fix header includes (TSLog.h -> Log.h)
- Remove pragma once from implementation files

Addresses memory leaks, crashes, and security vulnerabilities
identified during code review.

danielsreichenbach committed:
fix: resolve TypeScript compilation errors and type safety

- Fix TypeScript 5.6 compatibility issues
- Update tsconfig.json files for module resolution
- Fix import/export statements and module dependencies
- Resolve type definition conflicts and missing types
- Update VSCode launch and settings configurations
- Fix circular dependencies causing REPL issues
- Add missing type annotations and interfaces
- Update build scripts for TypeScript compatibility
- Fix deprecated code patterns and APIs

Makes codebase compile with TypeScript 5.6.

danielsreichenbach committed:
fix: moved include into platform guard

danielsreichenbach committed:
feat: migrate build system from tsc to SWC

Replace TypeScript compiler with SWC for faster compilation.
- Add .swcrc configuration for CommonJS and TypeScript
- Update build.js to use temp-scripts directory
- Fix module path resolution with symlinks
- Maintain TypeScript declaration generation
- Simplify addon path structure (remove nested directories)
- Fix circular dependencies by disabling decorator metadata
- Update all path references for new structure

Performance: ~10-20x faster compilation (300ms vs 3-6s)

Kaev commented:
Incredible work. Thank you! ❤️

danielsreichenbach committed:
fix: downgrade Ubuntu CI to clang-17 for sol2 compatibility

Ubuntu 24.04 ships with clang-18 which has a known regression that
breaks sol2 compilation. This causes template instantiation errors
in TSEventsLua.cpp with the error "address of overloaded function
'call' does not match required type 'int (lua_State *)'"

The issue is tracked in:
- ThePhD/sol2#1581 (original sol2 issue)
- tswow/tswow#855
- llvm/llvm-project#91362

Using clang-17 avoids this regression until sol2 is updated or
the fix is backported to clang-18.

danielsreichenbach committed:
fix: correct tswow_scripts path to prevent bin folder creation

The SWC migration incorrectly changed the tswow_scripts path from
'tswow-scripts' to 'bin/scripts', causing the build system to create
a bin folder in the project root instead of using the proper build
directory structure.

This restores the original path so that WotLK type definitions and
other build outputs are placed in the correct location within the
build directory.

danielsreichenbach commented:
For reference, the following project were used to test:

- <https://github.com/Duskhaven-Reforged/tswow-store>
- <https://github.com/tswow/tswow-tests>
- <https://github.com/magic62/magic-core>
- <https://github.com/magic62/shadows-of-azara>

danielsreichenbach committed:
feat: add duplicate realm ID detection at startup

Add validation to detect when multiple realms are configured with
the same ID, which previously caused the server to hang silently.

The check runs during realm initialization and:
- Collects all realm IDs across all modules
- Detects any duplicate IDs
- Displays a clear error message showing:
  - Which realm ID is duplicated
  - All modules using that ID (e.g. default.realm, magic-core.realm)
  - Instructions to fix by editing the realm.id file
- Exits gracefully with code 1

This prevents the confusing hang that occurred when multiple modules
had conflicting realm IDs and provides actionable guidance for
resolving the issue.

danielsreichenbach commented:
<img width="787" height="275" alt="Screenshot From 2025-08-01 18-58-50" src="https://github.com/user-attachments/assets/e8fc6a92-8c95-447d-b9be-41b692db0273" />

And because those projects have duplicate realm IDs I also added a quick feature to show that and exit instead of silently hanging until forever.

danielsreichenbach committed:
fix: simplify startup script to use start.js directly

Remove the intermediate start-tswow.js file and use start.js directly.
This is cleaner and less confusing for users who expect to see start.js
as the entry point.

Changes:
- Generate start.js directly instead of start-tswow.js
- Remove unnecessary wrapper script generation
- Update package.json template to use start.js in npm start command

The generated start.js now contains the wrapper code that ensures
--ipaths is set before loading the TSWoW runtime.

danielsreichenbach committed:
feat: add port conflict detection for realm servers

- Added getPort() method to Realm class that reads WorldServerPort from worldserver.conf
- Added port conflict detection in AuthServer.start() to check for duplicate ports
- System now warns when multiple realms use the same port (default: 8085)
- Warning message clearly indicates which realms have conflicts
- Suggests changing WorldServerPort in worldserver.conf to resolve conflicts

This prevents confusion when trying to start multiple realms that are configured
to use the same port, as only one realm can bind to a port at a time.

danielsreichenbach commented:
<img width="758" height="342" alt="image" src="https://github.com/user-attachments/assets/9f372b7a-43fe-403b-ad2c-db458c10b590" />

... and now it also warns if multiple realms use the same port. A little convenience.

danielsreichenbach committed:
fix: correct wow package.json module resolution paths

Changed main and types paths from "../data/index.js" to "./data/index.js"
to fix module resolution when the wow package is symlinked into node_modules.

The issue occurred because:
- bin/scripts/wow/ contains a symlink: data -> ../data
- node_modules/wow is a symlink to bin/scripts/wow/
- When Node.js resolved "../data/index.js" from the symlinked location,
  it incorrectly looked for node_modules/data/index.js

This fix resolves the "Cannot find module" error that prevented the
azara-core module from loading during datascript builds.

acidmanifesto commented:
> <img width="758" height="342" alt="image" src="https://github.com/user-attachments/assets/9f372b7a-43fe-403b-ad2c-db458c10b590" />
> 
> ... and now it also warns if multiple realms use the same port. A little convenience.

Now that right there is an awesome feature in itself.

danielsreichenbach committed:
fix: move wotlk TypeScript declarations to correct location

- Added rootDir to TypeScript config to properly handle relative paths
- This prevents creation of duplicate wow/wow/wotlk directory structure

danielsreichenbach committed:
fix: use platform-appropriate filename for blpconverter binary

- Changed hardcoded 'blpconverter.exe' to use isWindows() check
- Linux systems now get 'blpconverter' without .exe extension

danielsreichenbach committed:
fix: improve TSWoW startup reliability

- Add proper error handling to main() with catch block
- Fix npm install interference with authserver by using 'pipe' mode
- Prevent duplicate npm install calls that can cause startup issues
- Improve error messages when wow module installation fails
- Don't throw errors that prevent TSWoW from starting

These changes fix issues where the authserver would halt immediately
after initialization due to npm processes interfering with IO streams.

CharlieDelune commented:
I was going through, painstakingly making a lot of these same changes and ended up with failing builds anyway. Someone mentioned your PR, I pulled your fork, and ...

<img width="572" height="514" alt="image" src="https://github.com/user-attachments/assets/a2db641c-fbd7-426b-af6a-1bc5fff59482" />

Hooray!

danielsreichenbach committed:
fix: general TSWoW stability fixes

- Add SendCommand export to CommandActions for sending commands to services
- Add debug logging to AuthServer startup to detect immediate exits
- Fix npm install issues in Datascripts with better error handling
- Add debug logging throughout TSWoW initialization flow
- Add process lifecycle logging to track service startup/shutdown
- Detect when authserver halts unexpectedly

These changes help diagnose startup issues and improve overall reliability
without including any bot-specific functionality.

danielsreichenbach commented:
With the last commit all fixes I would want to see added for Linux are in. The build failures are due to the gitmodules not being set to updated hashes as mentioned in my initial post.

Blockost commented on tswow-scripts/util/Paths.ts:
```diff
@@ -672,6 +671,7 @@ export function BuildPaths(pathIn: string, tdb: string) {
         }),
 
         adtcreator: dir({
+            adt_creator_exe: file(`adt-creator${isWindows()?'.exe':''}`),
```
@danielsreichenbach Quick comment here related to paths for macOS: `Paths.ts` only seems to support either Windows or Linux builds. On macOS, shared libraries use `.dylib` file extension. 

I had to patch that on my side, but it'd be great if we could have a mechanism to also support macOS in here: https://github.com/tswow/tswow/pull/915/files#diff-936eb42d359b28f61dd756bc377bcf7e559be33fe871552b40114bbf7d610823R613

Blockost replied:
I'd be glad to help you on that. Looking at how paths are built, it could be great to have something a bit more flexible and robust in `CompilePath.ts` for example: https://github.com/tswow/tswow/blob/master/tswow-scripts/compile/CompilePaths.ts#L3-L9

danielsreichenbach replied:
> I'd be glad to help you on that. Looking at how paths are built, it could be great to have something a bit more flexible and robust

I will look at this today.

danielsreichenbach replied:
> I had to patch that on my side, but it'd be great if we could have a mechanism to also support macOS in here

How did you patch this? I can not test currently I just ordered a used Mac m1 for testing

Blockost replied:
I simply used `.dylib` for each lib:
https://github.com/Blockost/tswow/commit/074a960e86229bc6a3dfe5ec01d0d6563326ed05#diff-936eb42d359b28f61dd756bc377bcf7e559be33fe871552b40114bbf7d610823R635-R639

But having a way to distinguish between different platforms (maybe creating a helper function in `Platform.ts`) would be ideal

SirProdigle commented:
Got this working on Fedora Linux but required
```bash
export CC=clang-17
export CXX=clang++-17
export LD=/usr/bin/ld.bfd
``` 
as well as a **manual make/install of efsw** (probably a good idea to put a check at the start of the npm run build script for it)

There's some GNU only flags such as --no-undefined which ld.bfd is required to parse correctly, so that could be resolved by getting rid of those files from whatever make config has them.

also required an addition to the core CMakeLists.txt of 
```
install(TARGETS TracyClient
  ARCHIVE DESTINATION lib
  LIBRARY DESTINATION lib
  RUNTIME DESTINATION bin
)

```
anywhere after the 
```
FetchContent_Declare(
  tracy

```

I haven't figured out why this part is required yet, it seems to work on other linux distros but is having issue on my Fedora.

brndd commented:
Just for the record on Fedora (and any other distro for that matter) you probably want to use `CC=clang-17 CXX=clang++-17` because AFAIK the current revision of TrinityCore won't build with anything else.

Also, for EFSW you may have to pass -DBUILD_EFSW=ON to Cmake.

danielsreichenbach committed:
feat: add visibility to asset processing in datascripts build

This commit enhances the datascripts build process to show what's happening
with module assets during the build.

Key improvements:
- Always show basic asset linking information (module name, file count, patch target)
- Show detailed paths and file types when --debug flag is used
- Add logging to BLP conversion process (requires --build-blp flag)
- Add logging to taxi map processing

The asset flow is now visible: Module assets → Symlinks → MPQ patches → Client

Example output:
[datascripts] Linking 2 module asset folder(s) to client patches:
[datascripts]   - tswow-tests: 156 files → patch-4.MPQ
[datascripts]   - azara-core: 42 files → patch-5.MPQ

With --debug, additional details like paths and file types are shown.

danielsreichenbach committed:
fix: normalize MPQ internal paths to use backslashes consistently

This commit fixes inconsistent path separators in MPQ archive internal paths.
Previously, DBC files used hardcoded backslashes while asset and LUAXML files
used platform-dependent separators from Node.js path.relative().

Changes:
- Force backslashes for LUAXML internal paths: rel.replace(/\/g, '\\')
- Force backslashes for asset internal paths: relativeTo().replace(/\/g, '\\')

This ensures all internal MPQ paths use backslashes regardless of build platform,
matching Blizzard's MPQ format standard and maintaining consistency with DBC paths.

While StormLib auto-normalizes paths, this change ensures consistent MPQ contents
across platforms and better compatibility with other MPQ tools.

danielsreichenbach committed:
fix: ORM compatibility with MySQL 8.0.17+ display width changes

Fixes #916 where ORM tables were being rebuilt on every launch
when using MySQL 8.0.17 or later.

Problem:
- MySQL 8.0.17+ no longer reports display width for integer types
- Expected: int(10) unsigned, bigint(20), tinyint(3) unsigned
- Actual:   int unsigned, bigint, tinyint unsigned
- This caused ORM schema comparison to fail, triggering table rebuilds

Solution:
- Add normalizeColumnType() function to remove display width from both
  expected and actual column types before comparison
- Handles all integer types: tinyint, smallint, mediumint, int, bigint
- Applied to both primary key and regular column type comparisons

This ensures ORM tables are only rebuilt when actual schema changes occur,
not due to MySQL version differences in type reporting.

danielsreichenbach committed:
fix: convert WNode objects to strings before path normalization

Fixes TypeError: rel.replace is not a function error that occurred after
the path separator normalization changes.

Problem:
- node.relativeTo() returns a WNode object, not a string
- Calling .replace() on a WNode object throws TypeError
- This affected both LUAXML and asset path processing

Solution:
- Add .get() calls to convert WNode objects to strings before .replace()
- Fixed in both locations:
  - LUAXML: rel.get().replace(/\/g, '\\')
  - Assets: node.relativeTo(x.assets.path).get().replace(/\/g, '\\')

danielsreichenbach committed:
fix: use bundled node and quote paths for spaces

- Revert to using bundled node.exe on Windows instead of system node
- Add quotePath() helper to quote paths containing spaces
- Quote all executable and file paths passed as command arguments
- Fixes issues when TSWoW is installed in paths with spaces

This resolves the issue where the system node was being used
instead of the bundled node, and commands were failing when
paths contained spaces.

danielsreichenbach committed:
fix: add Windows support for symlink creation in build scripts

- Add platform detection for symlink creation
- Use mklink /J (directory junctions) on Windows instead of ln -sf
- Directory junctions don't require admin privileges on Windows
- Quote all paths in symlink commands to handle spaces
- Fix relative vs absolute path issues for symlink targets

This ensures module resolution works correctly on both Windows
and Unix systems during the build process.

danielsreichenbach committed:
refactor: use Node.js built-in fs.symlinkSync for symlinks

- Replace shell commands (ln -sf, mklink) with fs.symlinkSync
- Automatic platform detection (junction on Windows, dir on Unix)
- Better error handling and logging
- No shell command quoting issues
- More maintainable and portable code

This eliminates shell-specific commands and quoting issues,
making the symlink creation more robust and cross-platform.

danielsreichenbach committed:
fix: handle arrow functions in forEach correctly in typescript2cxx

- Add proper type checking before accessing CallExpression.arguments
- Prevent crash when arrow function parent is not a CallExpression
- Fixes issue where functions inside forEach were not being processed

The issue occurred when the TypeScript compiler tried to access
arguments on a parent node that wasn't actually a CallExpression,
which commonly happens with forEach callbacks.

danielsreichenbach committed:
fix: quote npx executable in TypeScript declaration generation

- Import NpxExecutable from runtime/Node for consistency
- Quote npx path to handle spaces in installation paths
- Ensures TypeScript declaration generation works with bundled node

This prevents the 'Option project cannot be mixed with source files'
error when paths contain spaces.

danielsreichenbach committed:
fix: resolve TypeScript declaration generation errors

- Add 'tswow' to TerminalCategory type definition
- Fix incorrect import path for __wotlkEvents module
- Remove error filtering to expose all TypeScript compilation issues

This eliminates all TypeScript declaration errors and ensures
proper type checking instead of hiding problems. The build
now generates TypeScript declarations successfully without errors.

danielsreichenbach committed:
fix: correct wow package location in build process

The wow package was being created in ipaths.bin.scripts instead of
ipaths.node_modules.wow where the runtime expects it. This caused
the error "Cannot find module 'node_modules/wow/data/index.js'"
when running datascripts build.

The fix moves the wow directory creation to the correct location
and updates symlink targets to point to the bin.scripts directory.

danielsreichenbach committed:
fix: correct relative symlink paths in wow package

The symlinks in node_modules/wow were using incorrect relative paths.
From node_modules/wow, the correct path to bin/scripts should be
../../bin/scripts, not ../bin/scripts.

This fixes the module resolution for wow/data/index.js.

danielsreichenbach committed:
fix: create wow package in both required locations

The TSWoW runtime expects the wow package to exist in two places:
1. node_modules/wow - for module resolution when running datascripts
2. bin/scripts/wow - as a local package that can be installed via npm

This change modifies Scripts.ts to create the wow package in both
locations with appropriate symlinks and package.json files:

- node_modules/wow has symlinks: ../../bin/scripts/data, ../../bin/scripts/util
- bin/scripts/wow has symlinks: ../data, ../util

Both directories get identical package.json and wotlk directory copies.
This allows both `require('wow/data')` and `npm i "bin/scripts/wow"`
to work correctly.

danielsreichenbach committed:
fix: use absolute paths for node_modules/wow symlinks

The relative symlink paths were not working correctly in the install
directory context. Changed to absolute paths to ensure symlinks resolve
properly when accessed from the install directory.

This fixes the "Cannot find module 'wow/data'" error, revealing the next
layer of the architectural issue: import paths that differ between build
and install contexts.

danielsreichenbach committed:
fix: correct import path for __wotlkEvents in install context

Changed import from '../wotlk/internal/__wotlkEvents' to
'../wow/wotlk/internal/__wotlkEvents' to match the install directory
structure where wotlk files are located within the wow package.

This resolves the fundamental architectural issue where import paths
differed between build and install contexts. The build process now
generates context-appropriate import paths for the runtime environment.

Fixes the "Cannot find module '../wotlk/internal/__wotlkEvents'" error
and allows datascripts to execute successfully.

tapmondev commented on tswow-scripts/compile/Scripts.ts:
```diff
@@ -25,75 +25,167 @@ import { bpaths, spaths } from './CompilePaths';
+        // Create symlinks for data and util directories so module resolution works correctly
+        const dataSymlink = mpath(wowDir, 'data');
+        if (!wfs.exists(dataSymlink)) {
+            wsys.exec(`ln -sf ../data ${dataSymlink}`, 'inherit');
```
this fails on windows

tapmondev replied:
Was using old wiki recommended node 20.18. fixed by updating to node 24!

danielsreichenbach replied:
Anything from NodeJS 22.x on should work.

ihm-tswow commented on tswow-scripts/compile/Scripts.ts:
```diff
@@ -25,75 +25,167 @@ import { bpaths, spaths } from './CompilePaths';
 
 export namespace Scripts {
     export async function build() {
```
(moved to line comment to allow proper discussion)

I don't think I agree with moving source script compilation to swc. When actively editing std or shell scripts, a common workflow is keeping the npm run build --interactive up to run the tsc watchers that generates both scripts and declarations.

Datascripts currently use swc because they can be massive, often do not need to generate declarations (unless referenced by another module, which uses tsc watchers instead) and because the weird package dependencies that datascript uses often causes tsc watchers to break, which confuses less technical users.

Source scripts rarely have these instability problems, and do require types to be generated (at least tswow-scripts/data does, which is the heaviest build). Since the current approach requires separately building declarations, tsc is still invoked and loses much of the performance benefit of using swc.

ihm-tswow commented on tswow-install/bin/scripts/data/Settings.d.ts:
```diff
@@ -0,0 +1,628 @@
+import { WDirectory } from "../util/FileTree";
```
(applies to all tswow-install files) these files should not be source-controlled, looks accidental (usually tswow-install is outside source directory, so i forgot to add it to .gitignore)

ihm-tswow reviewed (changes requested):
(not full reivew, i'll need to look more later)
